### ISSUE TITLE
Support values larger than `i64`

### ISSUE THREAD

**Issue by JustinLovinger** [nushell/nushell#8054]
### Related problem

I have a JSON file containing `u64` values. If I try to `open` this file and `get` certain `u64` fields, or manipulate the data in various other ways, Nushell throws error "can't convert value larger than i64 to i64 sized integer".

### Describe the solution you'd like

Nushell should support arbitrarily large integer values.

### Describe alternatives you've considered

Use `jq` instead of Nushell.

### Additional context and details

If Nushell only support up to `u64`, a similar issue could occur for `u128` or big integer types.

**Comment by sholderbach** [comment]
nushell did support big-ints for a while before version `0.59`. With the refactor of the underlying data model at the time it was removed for performance reasons and to ensure internally consistent handling of integer values as there would be several boundaries that would have to respect big-int types. We are certainly aware that the current `nu_protocol::Value` is not yet the be-all-end-all solution.

**Comment by repi** [comment]
we've ran into this as well, we have 128-bit GUIDs and in JSON output actually split it in 2 64-bit values for compatibility reasons but nushell does fail on them with `CantConvert { to_type: "i64 sized integer", from_type: "value larger than i64", span: Span { start: 118184,  end: 118194 }, help: None }`.

so at least natively supporting unsigned 64-bit integers would be great as feels quite common.

**Comment by sholderbach** [comment]
I think with the added difficulty of defining clear semantics for the interactions of unsigned and the existing signed types, I think adding a bigint representation internally is the more seamless solution for the problem of dealing with arbitrary incoming data (provided we use format parsers that support that).

Even though there might be benefits to having fixed length unsigned types for our `bytes` and `bits` operations, my gut instinct is "Go bigint or go home... err back to the drawing board"

**Comment by amtoine** [comment]
@repi 
do you have a simple JSON structure to try it out? :yum:

**Comment by repi** [comment]
@amtoine yes!

```json
{
    "hi": 12390093829892492485,
    "lo": 11667719501832895485
}
```

```sh
$ open guid.json                                                       09/14/2023 12:57:51 AM
╭────┬────────────────────────────────────────────────────────────────────────────────────────────────╮
│ hi │ CantConvert { to_type: "i64 sized integer", from_type: "value larger than i64", span: Span {   │
│    │ start: 119246, end: 119250 }, help: None }                                                     │
│ lo │ CantConvert { to_type: "i64 sized integer", from_type: "value larger than i64", span: Span {   │
│    │ start: 119246, end: 119250 }, help: None }                                                     │
╰────┴────────────────────────────────────────────────────────────────────────────────────────────────╯
```

**Comment by jntrnr** [comment]
@justinlovinger / @repi - I haven't seen this come up in ages.  Like the other folks said, we used support BigInteger and BigDecimal. They had expressive power, but were definitely slower than machine int and float. We're currently working on tuning the engine, so there's definitely a part of me hesitant to add more types or to return to using big number again.

Related: I had a look around the JSON spec/rfc because I couldn't remember that corner of JSON. As best as I can read, it seems like u64 is not supported in JSON:

https://datatracker.ietf.org/doc/html/rfc8259#section-6

In the RFC, the range given for ints to be compatible is [-(2**53)+1, (2**53)-1]. Having said that, I have no doubt there are variants of JSON that will happily support u64. Trying to support non-standard JSON is less motivating I think, but if we have good use cases, we could tackle those.

**Comment by repi** [comment]
suspected it was not standard JSON, and this is a data structure we ourselves are in control over at least so we can likely make it compliant.

so understandable to not support. probably are some bigint cases out there, but until then.

**Comment by justinlovinger** [comment]
> Related: I had a look around the JSON spec/rfc because I couldn't remember that corner of JSON. As best as I can read, it seems like u64 is not supported in JSON:

This is not just a problem for JSON. Nushell supports many serialization formats, and I suspect it will support more in the future. I would be surprised if none of those formats officially support values larger than `i64`.

Furthermore, supporting large integers is useful beyond serialization formats. Some mathematical operations, like exponentiation and factorial, can easily result is values larger than `i64`.

**Comment by amtoine** [comment]
just wanted to add that doing
```nushell
let data = {
    "hi": 12390093829892492485,
    "lo": 11667719501832895485
}
```
and then
```nushell
> $data
──┬───────────────────────
hi│12390093829892493312.00
lo│11667719501832894464.00
──┴───────────────────────
```
works just fine in the REPL.

**Comment by jntrnr** [comment]
@amtoine - that doesn't look correct. The number is changing.

**Comment by amtoine** [comment]
> @amtoine - that doesn't look correct. The number is changing.
>
ooh wow, i was fooled by the parsing working correctly in the REPL, but you are correct :scream:

**Comment by jntrnr** [comment]
@justinlovinger 

> Furthermore, supporting large integers is useful beyond serialization formats. Some mathematical operations, like exponentiation and factorial, can easily result is values larger than i64.

This is why I was hoping we could get some clear examples of the need. If we can, it'd be easier to design for them. An i64 is an enormous number. In practice, I'm not sure how often people would exceed it in their work. Do you have examples where it would happen?

**Comment by justinlovinger** [comment]
> Do you have examples where it would happen?

The situation that led to me making this issue was trying to examine an applications state. I had a large list of events from the application, it was using event-sourcing, which I exported from the application and opened using Nushell, to examine in an exploratory manner. I was trying to figure out how much space was used by various subsets of events. Some events included RNG state, which included max-value `u64`s. Those values caused exceptions when performing various operations, which interfered with examining the data in general, even though I was not trying to view RNG state.

P.S. I exported the data in JSON, but I could have exported it in any format supported by Serde.

**Comment by jntrnr** [comment]
@justinlovinger - for now I think the answer is to view the u64 (or other unsupported data formats) as binary and work with them like that. This doesn't allow for any kind of numeric manipulation, but would at least allow you to view them.

We may explore adding other number types, or possibly growing the default number type at some point, though we're pretty focused on getting 1.0 done atm, so it'll likely have to wait.

**Comment by mikeplus64** [comment]
Would be nice if this was added back. As noble a pursuit it is to have as high performance as possible, it's a shell + scripting language. Surely having great interop should be prioritised over performance?

**Comment by devyn** [comment]
Bignums probably make sense, as annoying as they are to use internally. We can try to just get really solid on fallible conversions to Rust fixed integers where we need them, for ergonomic reasons.

I ran into this again with the MessagePack format, which does natively support u64, and I decided on just throwing an error if we get one that's too big because `integer or binary` is not a very typically supported type combination. It would be nice to not have to really think about it.

**Comment by manveru** [comment]
Just wanted to add that i stumbled over this issue as well while trying to parse some `etcd` output, and the member ids it uses can be larger than i64. So for now I use an unholy combination of `jq` and `gawk` to convert them to hex for further processing.

```
http get --raw http://127.0.0.1:5001/members
  | jq -s '.[] | map({id: .id, name: .name})'
  | awk '{ if (match($0,/"id": ([0-9]+),/,m)) { printf "\"id\": \"0x%x\",", m[1] } else { print $0 }}'
  | from json
```

I never expected that Nushell would have a hard limit for Numbers. Isn't it possible to handle them like in Ruby, where they only get converted to Bignums when they exceed the size of i64?

**Comment by justinlovinger** [comment]
> Isn't it possible to handle them like in Ruby, where they only get converted to Bignums when they exceed the size of i64?

Ideally, that would be handled by the bignum implementation itself. There was work being done for that in `num-bigint`, <https://github.com/rust-num/num-bigint/pull/210>, but it appears to have stalled.

**Comment by manveru** [comment]
I think for the time being, the least Nushell could do is to make an error if it encounters numbers that are outside the supported range. Silent corruption of inputs is making me really paranoid.

As a good example, there's also already a check on exponentiation:
```
❯ : (2 ** 63)
Error: nu::shell::operator_overflow

  × Operator overflow.
   ╭─[entry #85:1:4]
 1 │ (2 ** 63)
   ·    ─┬
   ·     ╰── pow operation overflowed
   ╰────
  help: Consider using floating point values for increased range by promoting operand with 'into float'. Note: float has reduced precision!
```
However, these case right now fail silently and catch me off guard if I don't pay close attention:

```
❯ : 9223372036854775808
9223372036854776000
❯ : '9223372036854775808' | into int
9223372036854775807
❯ : 0xffffffffffffffff
-1
❯ : -9223372036854775808 | describe
int
❯ : -9223372036854775809 | describe
float
```

What I found especially funny is that `0xfffffffffffffffff` is interpreted as a command.

In the meantime, here's a `jq` snippet that converts all numbers in JSON to strings, to at least preserve the values. Hopefully this helps someone.

```
const filter = '
  def f: map_values(
    if type == "number" then tostring
    elif type == "object" then f
    elif type == "array" then f
    else .
    end
  );
  f
'
```
can be used like this:
```
something-that-outputs-json | jq $filter | from json
```

While this is less than ideal, at least I find it more reassuring that I have to explicitly handle each number I would like to use in operations and ensure it remains within an acceptable range.

That usually looks something like this:

```
const max = 9223372036854775807
const min = -9223372036854775808

def safe_add [a, b] {
  if ($a > 0 and $b > $max - $a) {
    error make {msg: "integer overflow", span: (metadata $a).span}
  } else if ($a < 0 and $b < $min - $a) {
    error make {msg: "integer underflow", span: (metadata $a).span}
  }
  $a + $b
}

safe_add $max 1
```

Long story short, I'm a big fan of BigDecimal and Rational number support in all languages I use if they don't have explicit type annotations and conversion sanity checks at every operation.
I understand speed is important for you, and I appreciate the focus on performance, but it should not come at the cost of correctness.
If I had more time and was better at Rust and math, I'd love to help out in the implementation, but unfortunately for now my rambling commentary will have to suffice as contribution to alleviating this issue.

**Comment by lizclipse** [comment]
A niche use-case I've run into this issue with is my [ULID plugin](https://github.com/lizclipse/nu_plugin_ulid), as ULIDs are just `u128`s with the random portion being 80 bits of that. When parsing out a ULID the timestamp is easy to convert into a Nushell value, however the random portion cannot be represented properly with the current `i64` limit. So far I've just been using a string since it's easy to view and move around, but proper support for larger ints would be quite useful here (even `u/i128` would work for this, but a more general solution would likely be better).
